README.md
`youngatlas` is a jupyter notebook python package extension. 


# The Story #


These days you call your friends to keep in touch. One time, I was catching up with an old colleague of mine and towards the tail end of our conversation we started talking about about how all of us have to do our part in dealing with this global COVID situation. He mentioned how his family built a COVID tracker dashboard. I can totally imagine how something like this could be useful to officials, organizations and communities. It was a great way to contribute.

I moved on to my daily/weekly chores like I usually do, feeling inspired.  Except this time there was this constant voice in my head that kept talking to me "you need to do a project and finish it and I will not leave you until you do" 

But I wasn't sure what to do. 

As part of my work, I usually spend hours on jupyter notebooks. 
A couple of weeks ago, almost towards the end of the day, exhausted, I was squinting hard on a piece of code on a notebook cell trying to make sense of it. There was too much distraction with my kids watching TV in the other room.
I couldn't focus on what I was doing and decided to take a break and join the kids. They were watching  `Avatar the last airbender` on Netflix. 

After watching the show for some time, I realized there was such a great deal of wisdom in the lines spoken by the characters and so much of their struggle is relevant to our current stressful times due to the pandemic.

After my break, I went back to my desk glaring at the same piece of code, and it seemed just obvious this time what the bug was. 
Right then, the idea for a notebook extension was born. How about an extension that plays some of the best quotes from the shows you enjoyed watching.
You can listen to just when you want to take a break or while you are having a down moment to remind you on things that truly matter. 

Or when a long running task on a cell such as training your neural net or a data transformation job is taking too long to finish, what if Iroh of Avatar can notify you when the long running task got completed with a pretty good advice and a reminder on enjoying and taking a tea break. 

So as a working prototype, I have quickly put together a notebook extension with just a few sound clips from some of my favourite characters from Avatar. Currently the extension has quotes from antihero zuko and my personal favourite zuko's uncle Iroh. And I have added a few more.


## How to Install ##


You can directly install after cloning from the source like so:
```
pip install git+https://github.com/addamit/youngatlas.git
```

Or git clone the repo first and install
```
git clone https://github.com/addamit/youngatlas.git
```


now cd to the cloned repo folder and run 
```
pip install setup.py
```


## How to run ##


After you have successfully installed the extension, you can run the extension in any jupyter notebook by typing the magic %youngatlas in any code cell, just before you start typing any code. 

See sample notebook in the example folder

Once the cell finishes execution, you will be showered with deep meaning wisdom audio clip (about 10-20 seconds) that might make a connection with you. Be sure to keep your computer/laptop sound button unmuted so you can hear the clip.


 # Contributing #

Send pull requests, open feature requests or DM me on twitter `@addamit` for any collaboration.

 
To build the package `python setup.py sdist bdist_wheel`

To install 
 `pip install --no-index --find-link=dist/ youngatlas==0.1`


For local development of packages:
pip install -e .

For manually enabling the extension,
To ensure the ipython extension is automatically defined in the IPython profile and loaded automatically when a notebook is launched, you can add it to the IPython profile config script. 

Open file `~/.ipython/profile_default/ipython_config.py` and add the name `youngatlas` in `c.InteractiveShellApp.extensions list`  

Here is a sample content for `~/.ipython/profile_default/ipython_config.py`

```
c.InteractiveShellApp.extensions = [
    'youngatlas',
]
```

And for the notebook extension, 
~/.jupyter/jupyter_notebook_config.json
{
  "NotebookApp": {
    "nbserver_extensions": {
      "yaserver": true
    }
  }
}
```
```
# Changes #
For changes, see the [change log](CHANGELOG.rst)

CHANGELOG.rst
Changelog
=========

0.0.1
-----

Changes:
~~~~~~~
- Initial Version. 

tests/conftest.py
"""
tests
"""

from IPython.testing.globalipapp import start_ipython
import pytest

@pytest.fixture(scope="session")
def session_ip():
	print ("session IPython")
	session_ip = start_ipython()
	session_ip.run_line_magic(magic_name="load_ext", line="yamagics")
	return session_ip


@pytest.fixture(scope="session")
def ip(session_ip):
	"""Verify the extension is automatically loaded."""
	status = session_ip.find_cell_magic("yamagics")
	print ("magics status: {}".format(status))
	session_ip.run_line_magic(magic_name="load_ext", line="yamagics")
	print ("Yielding IPython")
	yield session_ip

	# session_ip.run_line_magic(magic_name="unload_ext", line="yamagics")
	#session_ip.run_line_magic(magic_name="reset", line="-f")
	#ip.extension_manager.reload_extension('storemagic')

	

tests/test_yamagics.py
from IPython.utils.io import capture_output
from pytest import raises


# https://github.com/davidesarra/jupyter_spaces/blob/master/tests/test_magics.py

def test_magics_code_execution_simple(session_ip):
	"""Given code in a cell, the magics will always the code."""

	with capture_output() as captured:
			session_ip.run_cell(raw_cell=" x = 2+2; assert x == 4")
	print ("{}".format(captured.stdout))
	assert session_ip.user_global_ns["x"] == 4
	

def test_magics_code_execution_complex(session_ip):
	"""Given code cell with definitions, are run as well without issues."""
	
	with capture_output() as captured:
			session_ip.run_cell_magic("yamoment", "", "f = lambda x: x + 1; y = f(x=2); assert y == 3")
	print ("{}".format(captured.stdout))
	assert session_ip.user_global_ns["y"] == 3
	assert callable(session_ip.user_global_ns["f"])


def test_magics_code_execution_returns_random_clip(session_ip):
	"""Given multiple code cells, each run will randomly run a new quote."""

	clips = []
	with capture_output() as captured:
			session_ip.run_cell_magic("yamoment", "#MOMENTDEBUG", "2+2")
			session_ip.run_cell_magic("yamoment", "#MOMENTDEBUG", "2+2")
			session_ip.run_cell_magic("yamoment", "#MOMENTDEBUG", "2+2")
	print ("cell execution result :{}".format(captured.stdout))
	msg = captured.stdout
	msg = msg.replace("\n", "") # remove new line 
	#print (msg)
	clips_names = msg.replace("[MomentAudio]:",",") # remove the tag
	clips_names = clips_names[1:] # remove the front comma
	clips_list = clips_names.split(",") # split all quotes
	print (clips_list)
	# assert atleast there are two different quotes from a list of 3
	assert len(set(clips_list)) >= 2
		

def test_magics_code_execution_raises_exception(session_ip):
	"""Given code cell with syntax errors will also result in execution.
	Errors will be reported by the kernel and displayed appropriately."""

	with capture_output() as captured:
			with raises(SyntaxError):
					session_ip.run_cell_magic("yamoment", "", "2+%3")
	print ("cell execution result :{}".format(captured.stdout))




tests/test_yaserver.py
# import yaserver
import pytest
import requests

from notebook.serverextensions import toggle_serverextension_python
from notebook import nbextensions, serverextensions, extensions
from notebook.notebookapp import NotebookApp



# #TODO: fixture
# url = 'https://localhost:tree/{}'.format(yaserver.YASERVER_URI)

@pytest.fixture(scope="module")
def notebook():
	app = NotebookApp(nbserver_extensions={'yaserver': True})
	app.init_server_extensions()

def test_valid_endpoint_returns_audio_clip(notebook):
	"""Given a valid clip request, the server returns a valid audio clip."""
	assert False

def test_invalid_request(notebook):
	"""Requesting an invalid clip returns an empty response. """
	assert False


# def test_get_method_support():
# 	"""Server currently only supports GET."""
# 	assert False


# def test_audio_play_time():
# 	"""Given a valid GET request, server returns a valid clip less than 20 seconds play time."""
# 	assert False

src/yaserver/yaservercore.py
"""Notebook server extension that registers a custom endpoint 
with jupyter notebook server. 

This server extension returns a valid audio file when requested by the notebook UI.
"""
import os
from notebook.utils import url_path_join
from notebook.base.handlers import IPythonHandler

YASERVER_URI = '/yamoment'
QUOTES_FOLDER = 'quotes'
QUOTES_LOCATION = os.path.join(os.path.dirname(__file__), QUOTES_FOLDER)
YASERVER_ENDPOINT_URI_PATH = "{}/.*".format(YASERVER_URI)


class HelloWorldHandler(IPythonHandler):
    """Custom endpoint handler registers a GET method and returns a sound clip upon request."""

    def get(self):

        self.log.info("uri path: {}".format(self.request.uri))
        uri_file_name = self.request.uri.split('/')[-1]
        quote_file_name = uri_file_name.lower()

        if quote_file_name.endswith('mp3'):
            quote_file_path = "{}/{}".format('quotes', quote_file_name)
            # file_name = os.path.join(os.path.dirname(__file__), quote_file_path)
            file_name = os.path.join(QUOTES_LOCATION, quote_file_name)
            self.log.info("Returning content: {}".format(file_name))
            self.log.info("=======================")

            # xxx: check if file present. Also assuming no more than 1MB file for now.
            with open(file_name, "rb") as file_handle:
                data = file_handle.read(1024 * 1024)
                self.write(data)

            self.finish()
        else:
            self.finish()


def load_jupyter_server_extension(nb_app):
    '''
    Register a hello world handler.

    Based on https://github.com/Carreau/jupyter-book/blob/master/extensions/server_ext.py
    '''
    web_app = nb_app.web_app
    host_pattern = '.*$'
    route_pattern = url_path_join(
        web_app.settings['base_url'], YASERVER_ENDPOINT_URI_PATH)
    print("route_pattern  :{}".format(route_pattern))
    nb_app.log.info(
        "yaserver core enabled!. Route pattern: {}".format(route_pattern))
    web_app.add_handlers(host_pattern, [(route_pattern, HelloWorldHandler)])


src/yamagics/yamagics_core.py
"""
IPYthon Magics Extension to play audio without displaying the audio widget.  
"""
import os
import random 
import pathlib
import inspect

from typing import Optional
from IPython import get_ipython
from IPython.display import Audio, display
from IPython.core.magic import line_cell_magic, Magics, magics_class
from IPython.core.magic_arguments import argument, magic_arguments, parse_argstring



# https://stackoverflow.com/questions/61176900/jupyter-colab-play-sound-with-any-error-in-any-cell-play-sound-after-compl/61176901
from IPython.core.ultratb import AutoFormattedTB
# Catch any Exception, play error sound and re-raise the Exception
#-------------------------------------------------
# initialize the formatter for making the tracebacks into strings
itb = AutoFormattedTB(mode = 'Plain', tb_offset = 1)

SERVICE_ENDPOINT = '/hello'


from yaserver import QUOTES_LOCATION
all_choices = []
included_extensions = ['mp3']


class _InvisibleAudio(Audio):
    """
    An invisible (`display: none`) `Audio` element which removes itself when finished playing.
    Original sample based on https://stackoverflow.com/a/50648266.
    """

    def _repr_html_(self) -> str:
        audio = super()._repr_html_()
        audio = audio.replace(
            "<audio", '<audio onended="this.parentNode.removeChild(this)"'
        )        
        return f'<div style="display:none">{audio}</div>'
        # return f'<div">{audio}</div>'


@magics_class
class NotificationMagics(Magics):
    """
    IPython extension implementing the magic.
    """

    @magic_arguments()
    @argument(
        "-u",
        "--url",
        default="quote1.mp3",
        help="URL of audio file to play.",
    )
    @argument(
        "line_code",
        nargs="*",
        help="Other code on the line will be executed, unless this is called as a cell magic.",
    )
    @line_cell_magic
    def moment(self, line: str, cell: Optional[str] = None):
        args = parse_argstring(self.moment, line)        
        MOMENTDEBUG = False
        if line and line == '#MOMENTDEBUG':
            MOMENTDEBUG = True

        code = cell if cell else " ".join(args.line_code)
        try:
            ret = self.shell.ex(code)
        finally:
            
            quote_url = random.choice(all_choices)
            audio = _InvisibleAudio(url='{}/{}'.format(SERVICE_ENDPOINT, quote_url), autoplay=True)
            if MOMENTDEBUG:
                    print ("[MomentAudio]:{}".format(quote_url))
            display(audio)

        return ret

def load_ipython_extension(ipython):
    ipython.register_magics(NotificationMagics)    
    file_names = [fn for fn in os.listdir(QUOTES_LOCATION) if any(fn.endswith(ext) for ext in included_extensions)]
    all_choices.extend(file_names)
    # ipython.shell.set_custom_exc((Exception,), custom_exc)

#get_ipython().register_magics(NotificationMagics)

src/yamagics/yamagicscore.py
"""
IPYthon Magics Extension to play audio without displaying the audio widget.  
"""
from yaserver import QUOTES_LOCATION, YASERVER_URI
import os
import random
import pathlib
import inspect

from typing import Optional
from IPython import get_ipython
from IPython.display import Audio, display
from IPython.core.magic import line_cell_magic, Magics, magics_class
from IPython.core.magic_arguments import argument, magic_arguments, parse_argstring


# https://stackoverflow.com/questions/61176900/jupyter-colab-play-sound-with-any-error-in-any-cell-play-sound-after-compl/61176901
from IPython.core.ultratb import AutoFormattedTB
# Catch any Exception, play error sound and re-raise the Exception
# -------------------------------------------------
# initialize the formatter for making the tracebacks into strings
itb = AutoFormattedTB(mode='Plain', tb_offset=1)


all_choices = []
included_extensions = ['mp3']


class _InvisibleAudio(Audio):
    """
    An invisible (`display: none`) `Audio` element which removes itself when finished playing.
    Original sample based on https://stackoverflow.com/a/50648266.
    """

    def _repr_html_(self) -> str:
        audio = super()._repr_html_()
        audio = audio.replace(
            "<audio", '<audio onended="this.parentNode.removeChild(this)"'
        )
        return f'<div style="display:none">{audio}</div>'
        # return f'<div">{audio}</div>'


@magics_class
class NotificationMagics(Magics):
    """
    IPython extension implementing the magic.
    """

    @magic_arguments()
    @argument(
        "-u",
        "--url",
        default="quote1.mp3",
        help="URL of audio file to play.",
    )
    @argument(
        "line_code",
        nargs="*",
        help="Other code on the line will be executed, unless this is called as a cell magic.",
    )
    @line_cell_magic
    def yamoment(self, line: str, cell: Optional[str] = None):
        args = parse_argstring(self.yamoment, line)
        MOMENTDEBUG = False
        if line and line == '#MOMENTDEBUG':
            MOMENTDEBUG = True

        code = cell if cell else " ".join(args.line_code)
        try:
            ret = self.shell.ex(code)
        finally:

            quote_url = random.choice(all_choices)
            audio = _InvisibleAudio(
                url='{}/{}'.format(YASERVER_URI, quote_url), autoplay=True)
            if MOMENTDEBUG:
                print("[MomentAudio]:{}".format(quote_url))
            display(audio)

        return ret


def load_ipython_extension(ipython):
    ipython.register_magics(NotificationMagics)
    file_names = [fn for fn in os.listdir(QUOTES_LOCATION) if any(
        fn.endswith(ext) for ext in included_extensions)]
    all_choices.extend(file_names)
    # ipython.shell.set_custom_exc((Exception,), custom_exc)

# get_ipython().register_magics(NotificationMagics)
